File: windows/Marquee.Core.Tests/StatusToneTests.cs
using Marquee.Core.Models;

namespace Marquee.Core.Tests;

// The library-status colors, the same as the website's
// lib/library/status-tone.ts and the Mac's API.LibraryStatus.tone.

public sealed class StatusToneTests
{
    [Theory]
    [InlineData("owned", StatusTone.Owned)]
    [InlineData("tracked_downloading", StatusTone.Downloading)]
    [InlineData("tracked_monitored", StatusTone.Missing)]
    [InlineData("coming_soon", StatusTone.Soon)]
    [InlineData("untracked", StatusTone.Neutral)]
    [InlineData("something_newer", StatusTone.Neutral)]
    public void EveryStatusWearsItsOwnTone(string status, StatusTone tone)
    {
        Assert.Equal(tone, LibraryStatus.FromValue(status).Tone);
    }

    [Fact]
    public void NoTwoStatusesShareATone()
    {
        var tones = LibraryStatus.Known.Select(status => status.Tone).ToList();
        Assert.Equal(tones.Count, tones.Distinct().Count());
    }

    [Fact]
    public void OnlyTitlesInTheLibraryGetAPosterStrip()
    {
        foreach (var status in LibraryStatus.Known)
        {
            Assert.Equal(status.IsInLibrary, status.Tone.HasPosterStrip());
        }
    }

    [Fact]
    public void TheColorKeyExplainsEveryStatus()
    {
        foreach (var status in LibraryStatus.Known)
        {
            Assert.False(string.IsNullOrEmpty(status.Name));
            Assert.False(string.IsNullOrEmpty(status.Meaning));
        }
        Assert.Equal("In your library", LibraryStatus.Owned.Name);
        Assert.Equal("Missing", LibraryStatus.TrackedMonitored.Name);
    }
}

File: mac/MarqueeTests/StatusToneTests.swift
import XCTest
@testable import Marquee

/// The library-status colors, shared with the website's
/// lib/library/status-tone.ts: every status its own tone, and the badge,
/// poster strip and search pill all derived from it.
final class StatusToneTests: XCTestCase {
    func testEveryStatusHasItsOwnTone() {
        XCTAssertEqual(API.LibraryStatus.owned.tone, .owned)
        XCTAssertEqual(API.LibraryStatus.trackedDownloading.tone, .tracked)
        XCTAssertEqual(API.LibraryStatus.trackedMonitored.tone, .missing)
        XCTAssertEqual(API.LibraryStatus.comingSoon.tone, .soon)
        XCTAssertEqual(API.LibraryStatus.untracked.tone, .neutral)
        XCTAssertEqual(API.LibraryStatus.unknown("later").tone, .neutral)

        let tones = API.LibraryStatus.knownCases.map(\.tone)
        XCTAssertEqual(Set(tones).count, API.LibraryStatus.knownCases.count)
    }

    func testOnlyTitlesInTheLibraryGetAPosterStrip() {
        XCTAssertNotNil(Theme.statusStrip(.owned))
        XCTAssertNotNil(Theme.statusStrip(.trackedDownloading))
        XCTAssertNotNil(Theme.statusStrip(.trackedMonitored))
        XCTAssertNotNil(Theme.statusStrip(.comingSoon))
        XCTAssertNil(Theme.statusStrip(.untracked))
        XCTAssertNil(Theme.statusStrip(.unknown("later")))
    }

    func testStripBadgeAndSearchPillShareOneColor() {
        for status in API.LibraryStatus.knownCases where status != .untracked {
            let strip = Theme.statusStrip(status)
            XCTAssertEqual(strip, status.tone.palette?.foreground, status.rawValue)
            XCTAssertEqual(SuggestionKindPill.colors(for: status)?.foreground, strip, status.rawValue)
        }
        XCTAssertEqual(Theme.statusStrip(.trackedMonitored), Theme.missing)
        XCTAssertEqual(Theme.statusStrip(.comingSoon), Theme.soon)
    }

    func testComingSoonRequestsWearTheSoonTone() {
        XCTAssertEqual(API.RequestTone.comingSoon.badgeTone, .soon)
        XCTAssertEqual(API.RequestTone.downloading.badgeTone, .tracked)
        XCTAssertEqual(API.RequestTone.owned.badgeTone, .owned)
    }

    func testTheColorKeyExplainsEveryStatus() {
        for status in API.LibraryStatus.knownCases {
            XCTAssertFalse(status.name.isEmpty)
            XCTAssertFalse(status.meaning.isEmpty)
        }
        XCTAssertEqual(SuggestionKindPill.accessibilityText(kind: .movie, status: .trackedMonitored), "Movie · Missing")
        XCTAssertEqual(SuggestionKindPill.accessibilityText(kind: .tv, status: .unknown("later")), "TV")
    }
}
